include/vigra/threadpool.hxx

```diff
@@ -0,0 +1,328 @@
+            throw std::runtime_error("enqueue on stopped ThreadPool");
+
+        tasks.emplace([task](int tid){ (*task)(tid); });
+    }
+    condition.notify_one();
+    return res;
+}
+
+// add new work item to the pool
+template<class F>
+void ThreadPool::enqueue(F&& f)
+{
+    //using return_type = typename std::result_of<F()>::type;
+    
+    //std::future<return_type> res = task->get_future();
+    {
+        std::unique_lock<std::mutex> lock(queue_mutex);
+
+        // don't allow enqueueing after stopping the pool
+        if(stop)
+            throw std::runtime_error("enqueue on stopped ThreadPool");
+
+        tasks.emplace(f);
+    }
+    condition.notify_one();
+    //return res;
+}
+
+// the destructor joins all threads
+inline ThreadPool::~ThreadPool()
+{
+    {
+        std::unique_lock<std::mutex> lock(queue_mutex);
+        stop = true;
+    }
+    condition.notify_all();
+    for(std::thread &worker: workers)
+        worker.join();
+}
+
+
+
+template<class ITER, class F>
+void parallel_foreach_impl(
+    ThreadPool & pool,
+    const uint64_t nItems,                   
+    ITER iter, 
+    ITER end, 
+    F && f,
+    std::random_access_iterator_tag
+){
+    // typedef typename std::iterator_traits<ITER>::reference ReferenceType;
+    uint64_t workload = std::distance(iter, end);
+    const float workPerThread = float(workload)/pool.nThreads();
+    const uint64_t chunkedWorkPerThread = std::max(uint64_t(std::floor(workPerThread/3.0f+0.5f)), uint64_t(1));
+
+    for( ;iter<end; iter+=chunkedWorkPerThread){
+
+        // ITER localEnd = iter+chunkedWorkPerThread < end ? iter + chunkedWorkPerThread : end;
+        const size_t lc = std::min(workload, chunkedWorkPerThread);
+        workload-=lc;
+        pool.enqueue(
+            [&f, iter, lc]
+            (int id)
+            {
+                for(size_t i=0; i<lc; ++i)
+                    f(id, iter[i]);
+            }
+        );
+    }
+    pool.waitFinished();
+}
+
+
+
+template<class ITER, class F>
+void parallel_foreach_impl(
+    ThreadPool & pool,
+    const uint64_t nItems,                   
+    ITER iter, 
+    ITER end, 
+    F && f,
+    std::forward_iterator_tag
+){
+
+    // typedef typename std::iterator_traits<ITER>::reference ReferenceType;
+    uint64_t workload = nItems;
+    const float workPerThread = float(workload)/pool.nThreads();
+    const uint64_t chunkedWorkPerThread = std::max(uint64_t(std::round(workPerThread/3.0f)),1ul);
+
+    for(;;){
+
+        const size_t lc = std::min(chunkedWorkPerThread, workload);
+        workload -= lc;
+        pool.enqueue(
+            [&f, iter, lc]
+            (int id)
+            {
+                auto iterCopy = iter;
+                for(size_t i=0; i<lc; ++i){
+                    f(id, *iterCopy);
+                    ++iterCopy;
+                }
+            }
+        );
+        if(workload==0)
+            break;
+        std::advance(iter, lc);
+    }
+    pool.waitFinished();
+}
+
+
```

Comment:
We might want to add a `parallel_foreach_impl` overload for `std::input_iterator_tag`. It would just assign tasks to each thread in turn. Normally, this would be undesirable due to locality, but for input iterators, it actually makes sense -- why wait for inputs that haven't arrived before utilizing the next thread?